### PR TITLE
Issue 176 handle circular reference in schema

### LINKED ISSUE #176: Degrade gracefully on definitions with circular references
Currently the renderer aborts with the message

> Failed to load definition. too much recursion

Using [topological sorting](https://en.wikipedia.org/wiki/Cycle_(graph_theory)#Cycle_detection) or another algorithm, it should be possible to identify circular `$ref`s and possibly omit those sections of the definition only, allowing the remainder of the definition to be rendered (with warnings?).

### PR DESCRIPTION
**Pull request by quangkhoa**
Resolves #176 

Also did some refactoring, and changed `type` to `types` in `Property` component to reflect that the property is an array.

### CODE REVIEW

File: src/parser/open-api/oneOfResolver.js [hunk 1]
```diff
@@ -13,15 +13,20 @@ import toPath from 'lodash/toPath'
...
+      const item = obj[key]
 
       if (key === 'oneOf') {
         paths.push(currentPath)
-      } else if (typeof obj[key] === 'object' || Array.isArray(obj[key])) {
-        walk(obj[key], currentPath)
+      } else if (typeof item === 'object' || Array.isArray(item)) {
+        if (!checkedNodes.has(item)) {
+          checkedNodes.set(item, true)
+          walk(obj[key], currentPath)
```
Comment (brendo) [inline src/parser/open-api/oneOfResolver.js]: `obj[key]` can be `item` here too :)

File: src/parser/open-api/allOfResolver.js [hunk 2]
```diff
@@ -1,5 +1,7 @@
 import { clone } from '../../lib/clone'
 
+const checkedNodes = new Map()
```
Comment (brendo) [inline src/parser/open-api/allOfResolver.js]: Consider a [`Set`](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/Set) instead if you just want to track existence of items.

File: src/parser/open-api/oneOfResolver.js [hunk 3]
```diff
@@ -13,15 +13,20 @@ import toPath from 'lodash/toPath'
  */
 function getOneOfPaths (obj) {
   let paths = []
+  const checkedNodes = new Map()
```
Comment (brendo) [inline src/parser/open-api/oneOfResolver.js]: Same deal as above, it looks like we don't care about the value here, so consider a `Set`.

File: src/parser/open-api/schemaParser.js [hunk 4]
```diff
@@ -11,127 +12,164 @@ const literalTypes = ['string', 'integer', 'number', 'boolean']
...
+ *
+ * @param {string} name
+ *
+ * @return {Object}
+ */
+function getUIPropertyForCircularReference (name) {
+  const visitedNodes = Array.from(checkedNodes.values()).join(', ')
+  const uiProperty = getBaseUIProperty('CIRCULAR REFERENCE', {})
+
+  uiProperty.description = `Circular reference found in **${name}**. Visited nodes: ${visitedNodes}`
```
Comment (brendo) [inline src/parser/open-api/schemaParser.js]: I don't think `visitedNodes` adds any value to the user.

eg. 

```
Circular reference found in **categories**. Visited nodes: added, elements, programme, images, initial_children, images, labels, master_brand, titles, next_broadcast, related_links, suppliers, synopses, labels, suppliers, synopses, atoz_programmes, broadcasts, channel, elements, duration, categories.
```

How does the list of nodes help me know what this object should be?

Comment (brendo) [reply src/parser/open-api/schemaParser.js]: I also find the "found in" part a little confusing. It's more saying, "I have seen this schema before in "categories"".

**Review by brendo — CHANGES_REQUESTED** [review]

**Comment by quangkhoa** [comment]
@brendo I've redone the solution. It should be simpler now.

**Review by brendo — APPROVED** [review]

**Pull request merged**